### PR TITLE
[SYCL][NFC] Remove device code compilation from unit tests

### PR DESCRIPTION
**Pull request by sergey-semenov**
Additionally, fix a couple of compilation warnings, a double free error
and a PiArray bug where MEntries might become invalidated due to
reallocation of MMockEntries.

### CODE REVIEW

File: sycl/cmake/modules/AddSYCLUnitTest.cmake [hunk 1]
```diff
@@ -46,58 +46,3 @@ macro(add_sycl_unittest test_dirname link_variant)
     )
   endif()
 endmacro()
-
-macro(add_sycl_unittest_with_device test_dirname link_variant)
```
Comment (s-kanaev) [inline sycl/cmake/modules/AddSYCLUnitTest.cmake:50]: I think it's still used in some `kernel-and-program` tests.

**Review by s-kanaev — DISMISSED** [review]
LGTM

File: sycl/cmake/modules/AddSYCLUnitTest.cmake (hunk 1 above)
Comment (sergey-semenov) [reply sycl/cmake/modules/AddSYCLUnitTest.cmake:50]: Not anymore, these are the last ones.

**Review by s-kanaev — APPROVED** [review]
LGTM

**Pull request merged**